### PR TITLE
Add in requirements.json creation feature

### PR DESCRIPTION
**Pull request by smlindauer**

### CODE REVIEW

File: src/sasctl/pzmm/write_json_files.py [hunk 1]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+        In the case of using this function outside the model development environment, it is
+        recommended to the user that they adjust the requirements.json file's package versions
+        to match the model development environment.
+
+        Parameters
+        ----------
+        json_path : str, optional
+            The path to a Python project, by default the current working directory.
+
+        Yields
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:1010]: Does not yield anything.

Currently returns None - is there a use-case for returning the JSON string (or dict) or do results always need to be written to disk?

File: src/sasctl/pzmm/write_json_files.py [hunk 2]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+            List of Python packages.
+
+        Returns
+        -------
+        list
+            Nested list of Python package_name names and found versions.
+        """
+
+        def package_not_found_output(package_name, package_versions):
+            print(
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:1078]: Do not print.  This should either be a warning or a log message.  Using print() prevents the user from controlling if/when such messages are suppressed.

File: src/sasctl/pzmm/write_json_files.py [hunk 3]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+            package_versions.append([package_name, None])
+            return package_versions
+
+        package_and_version = []
+        import importlib
+
+        for package in package_list:
+            try:
+                name = importlib.import_module(package)
+                try:
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:1093]: Not necessary, but might be more succinct to do something like:
`version = getattr(package, '__version__', None) or getattr(package, 'version', None) or getattr(package, 'VERSION', None)`

File: src/sasctl/utils/functionInspector.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python
```
Comment (jlwalke2) [inline src/sasctl/utils/functionInspector.py:1]: Per PEP8 module names should be lower case and include underscores if necessary.
[https://peps.python.org/pep-0008/#package-and-module-names](https://peps.python.org/pep-0008/#package-and-module-names)

File: src/sasctl/utils/functionInspector.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+# encoding: utf-8
+#
+# Copyright © 2022, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import sys
+
+
+def findPackageBytecodes(inst):
```
Comment (jlwalke2) [inline src/sasctl/utils/functionInspector.py:10]: Public function should have a docstring

File: src/sasctl/utils/functionInspector.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        if inst.argval in globals() and type(globals()[inst.argval]) == ModuleType:
+            return ("modules", [globals()[inst.argval]])
+        else:
+            return None
+    if "LOAD_" in inst.opname and type(inst.argval) in [CodeType, FunctionType]:
+        return ("code", inst.argval)
+    return None
+
+
+def getFunctionPackages(func):
```
Comment (jlwalke2) [inline src/sasctl/utils/functionInspector.py:35]: Public function should have a docstring

File: src/sasctl/utils/functionInspector.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                    elif inspect.iscode(bRef):
+                        codeworkList.append(bRef)
+        except TypeError:
+            continue
+    result = list(mods)
+    result.sort()
+    return result
+
+
+def findDependencies():
```
Comment (jlwalke2) [inline src/sasctl/utils/functionInspector.py:148]: Public function should have a docstring

File: src/sasctl/utils/functionInspector.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+# encoding: utf-8
+#
+# Copyright © 2022, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import sys
+
+
+def findPackageBytecodes(inst):
+    from types import CodeType, FunctionType, ModuleType
```
Comment (jlwalke2) [inline src/sasctl/utils/functionInspector.py:11]: move all the imports to top of file (they're standard packages that are always available and are always going to end up being loaded anyways).

File: src/sasctl/pzmm/write_json_files.py (hunk 1 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1010]: I could definitely see a use case for determining the development environment in the form of json or dict return. Is create_requirements_json() a broad enough function name to still work for that situation?

Comment (jlwalke2) [reply src/sasctl/pzmm/write_json_files.py:1010]: sounds good to me

File: src/sasctl/pzmm/write_json_files.py [hunk 9]
```diff
@@ -51,7 +75,12 @@ def writeVarJSON(cls, inputData, isInput=True, jPath=Path.cwd()):
                     else:
                         isStr = False
 
-                if isStr:
+            if isStr:
+                outputLevel = "nominal"
+                outputType = "string"
+                outputLength = predict.str.len().max()
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:81]: outputLevel, outputType, and outputLength appear to never actually be used.  Also, what is `predict` - I do not see it defined or passed as a parameter.

File: src/sasctl/pzmm/write_json_files.py [hunk 10]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
             conversion = 1
 
         return conversion
+
+    @classmethod
+    def create_requirements_json(cls, json_path=Path.cwd()):
+        """
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:988]: nitpick: single-line short description followed by an extended summary, per Numpydoc style guide: [https://numpydoc.readthedocs.io/en/latest/format.html#short-summary](https://numpydoc.readthedocs.io/en/latest/format.html#short-summary)

Multiple methods defined in te class have this (minor) problem

File: src/sasctl/pzmm/write_json_files.py [hunk 11]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+                            )
+            except ModuleNotFoundError:
+                package_and_version = package_not_found_output(
+                    package, package_and_version
+                )
+
+        return package_and_version
+
+    @classmethod
+    def get_code_dependencies(cls, json_path=Path.cwd()):
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:1113]: Would not call this `json_path`.  There's no JSON involved in this method, nor does the docstring reference JSON.

**Review by jlwalke2 — CHANGES_REQUESTED** [review]

File: src/sasctl/pzmm/write_json_files.py (hunk 1 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1010]: Committed changes (unpushed currently) that handle the json writing in a cleaner, more pythonic way and included a return of the list of dicts that represents the information in the requirements.json file.

File: src/sasctl/pzmm/write_json_files.py (hunk 2 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1078]: Converted to warnings.warn() here and at another location within write_json_files.py, as well as removed redundant try except block also attempting to print a warning.

File: src/sasctl/pzmm/write_json_files.py [hunk 12]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+                f"Warning: Package {package_name} was not found in the local environment. Either {package_name} is not "
+                f"a valid Python package, or the package is not present in this environment. The requirements.json file"
+                f" will include a commented out version of the pip installation command at the bottom of the file. "
+                f"Please review the file and verify that the package exists and input the version needed."
+            )
+            package_versions.append([package_name, None])
+            return package_versions
+
+        package_and_version = []
+        import importlib
```
[the remark this answers is not in the document]
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1088]: Done.

File: src/sasctl/utils/functionInspector.py (hunk 4 above)
Comment (smlindauer) [reply src/sasctl/utils/functionInspector.py:1]: Definitely outdated and should not have been included in the git commits. Removing it and resolving any other comments based on functionInspector.py.

File: src/sasctl/pzmm/write_json_files.py (hunk 9 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:81]: Something changed here that should not have. I'm going to parse through this section and see what is going on and where things broke down. Likely just revert it back to a prior state.

File: src/sasctl/pzmm/write_json_files.py (hunk 10 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:988]: Done.

File: src/sasctl/pzmm/write_json_files.py (hunk 11 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1113]: I think model_path probably makes more sense then, as the aim is to get the user pointing at the directory their model is in.

File: src/sasctl/pzmm/write_json_files.py [hunk 13]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+            File location for the output JSON file. Default is the current
+            working directory.
+
+        Returns
+        -------
+        list
+            List of found package dependencies.
+        """
+        file_names = []
+        file_names.extend(sorted(Path(json_path).glob("*.py")))
```
[the remark this answers is not in the document]
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1131]: Fixed.

File: src/sasctl/pzmm/write_json_files.py [hunk 14]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+        file_path : string or Path
+            File location for the Python file to be parsed.
+        Returns
+        -------
+        list
+            List of found package dependencies.
+        """
+        import ast
+
+        file_text = open(file_path).read()
```
[the remark this answers is not in the document]
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1158]: Fixed.

File: src/sasctl/pzmm/write_json_files.py [hunk 15]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+        pickle_folder : str
+            File location for the input pickle file. Default is the current working directory.
+        Returns
+        -------
+        list
+            A list of pickle files.
+        """
+
+        file_names = []
+        file_names.extend(sorted(Path(pickle_folder).glob("*.pickle")))
```
[the remark this answers is not in the document]
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1199]: Hmm. I don't believe path's glob supports complex regex, but I could just drop a list comprehension in instead and return that. Something like:
`[p for p in Path(pickle_folder).iterdir() if p.suffix in [".pickle", ".pkl"]]`

File: src/sasctl/pzmm/write_json_files.py [hunk 16]
```diff
@@ -950,3 +982,629 @@ def convertDataRole(self, dataRole):
...
+    @classmethod
+    def get_pickle_file(cls, pickle_folder=Path.cwd()):
+        """
+        Given a file path, retrieve the pickle file(s).
+
+        Parameters
+        ----------
+        pickle_folder : str
+            File location for the input pickle file. Default is the current working directory.
+        Returns
```
[the remark this answers is not in the document]
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:1192]: Fixed.

File: src/sasctl/pzmm/write_json_files.py [hunk 17]
```diff
@@ -10,12 +10,17 @@
 import math
 import pickle
 import pickletools
+import warnings
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:13]: imports should be grouped and ordered according to [PEP8](https://peps.python.org/pep-0008/#imports)

File: src/sasctl/pzmm/write_json_files.py [hunk 18]
```diff
@@ -38,25 +43,24 @@ class JSONFiles:
...
+            predictor and prediction variables (ambiguously named "predict"). Providing a list of dict objects signals
+            that the model files are being created from an MLFlow model.
+        isInput : bool
+            Boolean flag to check if generating the input or output variable JSON.
         jPath : string, optional
-            File location for the output JSON file. Default is the current
-            working directory.
+            File location for the output JSON file. Default is the current working directory.
+
+        Yields
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:60]: Take out the Yields section.  The function isn't a generator and never uses `yield`.  It doesn't `return` anything either (besides None) so you can't include the JSON file info in the Returns section either.

See [Numpydoc style guide](https://numpydoc.readthedocs.io/en/latest/format.html#yields)

Best place to document that it writes a file is probably the extended summary or possibly notes section.

File: src/sasctl/pzmm/write_json_files.py [hunk 19]
```diff
@@ -181,22 +181,20 @@ def writeModelPropertiesJSON(
...
-            working directory.
+            Location for the output JSON file. The default is the current working directory.
         modeler : string, optional
-            The modeler name to be displayed in the model properties. The
-            default value is None.
+            The modeler name to be displayed in the model properties. The default value is None.
 
         Yields
-        ---------------
+        ------
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:189]: Same issue as with writeVarJSON() docstring, the Yields section is inappropriate here.

File: src/sasctl/pzmm/write_json_files.py [hunk 20]
```diff
@@ -273,20 +271,18 @@ def writeFileMetadataJSON(cls, modelPrefix, jPath=Path.cwd(), isH2OModel=False):
...
         isH2OModel : boolean, optional
-            Sets whether the model metadata is associated with an H2O.ai model.
-            If set as True, the MOJO model file will be set as a score resource.
-            The default value is False.
+            Sets whether the model metadata is associated with an H2O.ai model. If set as True, the MOJO model file
+            will be set as a score resource. The default value is False.
 
         Yields
-        ---------------
+        ------
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:285]: Same issue as with writeVarJSON() docstring, the Yields section is inappropriate here.

File: src/sasctl/pzmm/write_json_files.py [hunk 21]
```diff
@@ -355,24 +352,20 @@ def writeBaseFitStat(
...
-            a sample parameter call would be ('NObs', 'Sum of Frequencies',
-            3488, 'TRAIN'). Variable dataRole is typically either TRAIN,
-            TEST, or VALIDATE or 1, 2, 3 respectively. The default value is None.
+            Input parameter tuples in the form of (parameterName, parameterLabel, parameterValue, dataRole). For
+            example, a sample parameter call would be ('NObs', 'Sum of Frequencies', 3488, 'TRAIN'). Variable dataRole
+            is typically either TRAIN, TEST, or VALIDATE or 1, 2, 3 respectively. The default value is None.
 
         Yields
-        ---------------
+        ------
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:368]: Same issue as with writeVarJSON() docstring, the Yields section is inappropriate here.

File: src/sasctl/pzmm/write_json_files.py [hunk 22]
```diff
@@ -499,7 +481,7 @@ def calculateFitStat(
             working directory.
 
         Yields
-        ---------------
+        ------
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:484]: Same issue as with writeVarJSON() docstring, the Yields section is inappropriate here.

File: src/sasctl/pzmm/write_json_files.py [hunk 23]
```diff
@@ -639,19 +616,19 @@ def generateROCLiftStat(
...
-            Dataframe, array, or list of the test data set, including both
-            the actual values and the calculated probabilities. The default value is None.
+            Dataframe, array, or list of the test data set, including both the actual values and the calculated
+            probabilities. The default value is None.
         jPath : string, optional
-            Location for the output JSON file. The default value is the current
-            working directory.
+            Location for the output JSON file. The default value is the current working directory.
 
         Yields
```
Comment (jlwalke2) [inline src/sasctl/pzmm/write_json_files.py:633]: Same issue as with writeVarJSON() docstring, the Yields section is inappropriate here.

**Review by jlwalke2 — CHANGES_REQUESTED** [review]
Looks good, with small changes needed.  Import order is nitpicky but Yields is actually wrong.  I did not add comments to every Yields section in the doc, but every docstring with a Yields section should be fixed - except for `flatten()` which uses it correctly.

File: src/sasctl/pzmm/write_json_files.py (hunk 17 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py:12]: Resolved.

File: src/sasctl/pzmm/write_json_files.py (hunk 18 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py]: Resolved.

File: src/sasctl/pzmm/write_json_files.py (hunk 19 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py]: Resolved.

File: src/sasctl/pzmm/write_json_files.py (hunk 20 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py]: Resolved.

File: src/sasctl/pzmm/write_json_files.py (hunk 21 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py]: Resolved.

File: src/sasctl/pzmm/write_json_files.py (hunk 22 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py]: Resolved.

File: src/sasctl/pzmm/write_json_files.py (hunk 23 above)
Comment (smlindauer) [reply src/sasctl/pzmm/write_json_files.py]: Resolved.

**Review by jlwalke2 — APPROVED** [review]

**Pull request merged**